pkg/envtest/crd.go

```diff
@@ -0,0 +1,200 @@
+	if err := readCRDFiles(&options); err != nil {
+		return options.CRDs, err
+	}
+
+	// Create the CRDs in the apiserver
+	if err := CreateCRDs(config, options.CRDs); err != nil {
+		return options.CRDs, err
+	}
+
+	// Wait for the CRDs to appear as Resources in the apiserver
+	if err := WaitForCRDs(config, options.CRDs, options); err != nil {
+		return options.CRDs, err
+	}
+
+	return options.CRDs, nil
+}
+
+// readCRDFiles reads the directories of CRDs in options.Paths and adds the CRD structs to options.CRDs
+func readCRDFiles(options *CRDOptions) error {
+	if len(options.Paths) > 0 {
+		for _, path := range options.Paths {
+			new, err := readCRDs(path)
+			if err != nil {
+				return err
+			}
+			options.CRDs = append(options.CRDs, new...)
+		}
+	}
+	return nil
+}
+
+// defaultCRDOptions sets the default values for CRDs
+func defaultCRDOptions(o *CRDOptions) {
+	if o.maxTime == 0 {
+		o.maxTime = defaultMaxWait
+	}
+	if o.pollInterval == 0 {
+		o.pollInterval = defaultPollInterval
+	}
+}
+
+// WaitForCRDs waits for the CRDs to appear in discovery
+func WaitForCRDs(config *rest.Config, crds []*apiextensionsv1beta1.CustomResourceDefinition, options CRDOptions) error {
+	// Add each CRD to a map of GroupVersion to Resource
+	waitingFor := map[schema.GroupVersion]*sets.String{}
+	for _, crd := range crds {
+		gv := schema.GroupVersion{Group: crd.Spec.Group, Version: crd.Spec.Version}
+		if _, found := waitingFor[gv]; !found {
+			// Initialize the set
+			waitingFor[gv] = &sets.String{}
+		}
+		// Add the Resource
+		waitingFor[gv].Insert(crd.Spec.Names.Plural)
+	}
+	// Poll until all resources are found
+	return wait.PollImmediate(options.pollInterval, options.maxTime, func() (bool, error) {
+		// Create a new clientset to avoid any client caching of discovery
+		cs, err := clientset.NewForConfig(config)
+		if err != nil {
+			return false, err
+		}
+
+		allFound := true
+		for gv, resources := range waitingFor {
+			// All resources found, do nothing
+			if resources.Len() == 0 {
+				delete(waitingFor, gv)
+				continue
+			}
+
+			// Get the Resources for this GroupVersion
+			resourceList, err := cs.Discovery().ServerResourcesForGroupVersion(gv.Group + "/" + gv.Version)
+			if err != nil {
+				return false, nil
+			}
+
+			// Remove each found resource from the resources set that we are waiting for
+			for _, resource := range resourceList.APIResources {
+				resources.Delete(resource.Name)
+			}
+
+			// Still waiting on some resources in this group version
+			if resources.Len() != 0 {
+				allFound = false
+			}
+		}
+		return allFound, nil
+	})
+}
+
+// CreateCRDs creates the CRDs
+func CreateCRDs(config *rest.Config, crds []*apiextensionsv1beta1.CustomResourceDefinition) error {
+	cs, err := clientset.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+
+	// Create each CRD
+	for _, crd := range crds {
+		if _, err := cs.ApiextensionsV1beta1().CustomResourceDefinitions().Create(crd); err != nil {
```

Comment:
wondering if we can use `pkg/client` do this ?

Reply:
We could, but i need this for discovery anyway...